src/v/cluster/producer_state.cc

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "producer_state.h"
+
+#include "cluster/logger.h"
+#include "cluster/producer_state_manager.h"
+#include "vassert.h"
+
+namespace cluster {
+
+std::optional<request_ptr> requests::last_request() const {
+    if (!_inflight_requests.empty()) {
```

Comment:
do we have a guarantee that `_inflight_requests.back() < _finished_requests.back()`? we wipe _inflight_requests on term change so it seems that there is a chance of it going backwards

Reply:
you got it right, the questing is if _inflight_requests can contain leftovers from previous term while finished_requests is already updated with new term; it looks like there is protection in `try_emplace` covering this case; so if it's the only place where were may insert new term's request feel free to resolve it